problem: unsecured websocket is rarely used in production

somdoron opened the pull request:
Solution: support websocket with tls (wss)

- [ ] Support CMake (added to #3581)
- [x] Support self-signed certificates
- [x] Move credentials to connecter and listener
- [ ] Validate certificates when set with setsockopt (added to #3581)
- [x] Load certificate
- [x] Change the option to enable-tls

bluca commented:
Nice stuff - any chance NSS or GnuTLS could be used as alternative to the vendorized library? For all the usual reasons :-)

somdoron commented:
Is the license of gnuTLS compatible?

On Mon, Sep 30, 2019, 14:00 Luca Boccassi <notifications@github.com> wrote:

> Nice stuff - any chance NSS or GnuTLS could be used as alternative to the
> vendorized library? For all the usual reasons :-)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/zeromq/libzmq/pull/3695?email_source=notifications&email_token=AAUW75XNJ7LKQL2RD3AUTZ3QMHL55A5CNFSM4I3S4KZKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD75IMBA#issuecomment-536512004>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAUW75W53DC5XBDGFWGGIVTQMHL55ANCNFSM4I3S4KZA>
> .
>

bluca commented:
Yes, it's LGPL2. Does NSS have the required functionality? Given we already have integration for it

somdoron commented:
I prefer gnuTLS as the docs are much better and no need for configuration files or database, as much as I can understand from the NSS docs and the nss_init method.

somdoron commented:
Switched to gnutls, gnutls also support sha1, so we can remove the dependency on NSS.

bluca commented:
Nice! There's a merge conflict

bluca commented on include/zmq.h:
```diff
@@ -670,6 +670,12 @@ ZMQ_EXPORT void zmq_threadclose (void *thread_);
 #define ZMQ_SOCKS_PASSWORD 100
 #define ZMQ_IN_BATCH_SIZE 101
 #define ZMQ_OUT_BATCH_SIZE 102
+#define ZMQ_TLS_KEY_PEM 103
```
if these are only for WS, perhaps it's worth naming ZMQ_WS_...? To avoid confusion

somdoron replied:
I will change it (to WSS), actually adding TLS to zmtp will be simple now and would make the name relevant, but if we do we can change them then and obsolete the previous names.

somdoron replied:
fixed

somdoron commented:
This is ready to be merged, some comments:
* I had to add another build to travis, because a minimum version of gnutls is required, which only available in xenial
* TLS cannot be enabled with CMake, I added to #3581 
* Some minor fixes are needed, I added them to #3581

bluca commented:
aren't we using xenial everywhere by now? anything earlier is deprecated

somdoron commented:
No... still trusty

bluca commented:
Looks like Windows builds are all failing, could you please have a look?

bluca commented on src/session_base.cpp:
```diff
@@ -114,8 +114,11 @@ zmq::session_base_t::session_base_t (class io_thread_t *io_thread_,
+    _wss_hostname (NULL)
 {
+    if (options_.wss_hostname.length () > 0)
+        _wss_hostname = strdup (options_.wss_hostname.c_str ());
```
```
1>..\libzmq\src\session_base.cpp(121) : error C2220: warning treated as error - no 'object' file generated
1>..\libzmq\src\session_base.cpp(121) : warning C4996: 'strdup': The POSIX name for this item is deprecated. Instead, use the ISO C++ conformant name: _strdup. See online help for details.
1>        C:\Program Files (x86)\Microsoft Visual Studio 9.0\VC\include\string.h(207) : see declaration of 'strdup'
1>select.cpp
2>server.cpp
2>..\libzmq\src\session_base.cpp(121) : error C2220: warning treated as error - no 'object' file generated
2>..\libzmq\src\session_base.cpp(121) : warning C4996: 'strdup': The POSIX name for this item is deprecated. Instead, use the ISO C++ conformant name: _strdup. See online help for details.
2>        C:\Program Files (x86)\Microsoft Visual Studio 9.0\VC\include\string.h(207) : see declaration of 'strdup
```

somdoron replied:
fixed.
